File: src/python/t135_Candy.py
def candy(self, ratings):
    n = len(ratings)
    result = 1
    up, down, peak = 0, 0, 0
    for i in range(1, n):
        if ratings[i - 1] < ratings[i]:
            up = up + 1
            peak = up
            down = 0
            result += 1 + peak
        elif ratings[i - 1] == ratings[i]:
            peak = up = down = 0
            result = result + 1
        else:
            up = 0
            down = down + 1
            result += down + 1 if peak < down else down
    return result

File: src/t135_Candy.java
public class t135_Candy {
    public int candy(int[] ratings) {
        int n = ratings.length;
        int result = 1;
        int up = 0;
        int down = 0;
        int peak = 0;
        for (int i = 1; i < n; i++) {
            if (ratings[i - 1] < ratings[i]) {
                peak = ++up;
                down = 0;
                result += 1 + peak;
            } else if (ratings[i - 1] == ratings[i]) {
                peak = up = down = 0;
                result++;
            } else {
                up = 0;
                down++;
                result += (peak < down) ? down + 1 : down;
            }
        }
        return result;
    }
}
